src/cryptonote_core/service_node_list.cpp

```diff
@@ -2743,21 +2744,44 @@ namespace service_nodes
     return true;
   }
 
-  static crypto::hash hash_uptime_proof(const cryptonote::NOTIFY_UPTIME_PROOF::request &proof, uint8_t hf_version)
+  std::string service_node_list::serialize_uptime_proof(const cryptonote::NOTIFY_UPTIME_PROOF::request &proof) const
   {
-    auto buf = tools::memcpy_le(proof.pubkey.data, proof.timestamp, proof.public_ip, proof.storage_port, proof.pubkey_ed25519.data, proof.qnet_port, proof.storage_lmq_port);
-    size_t buf_size = buf.size();
-
-    if (hf_version < cryptonote::network_version_15_lns) // TODO - can be removed post-HF15
-      buf_size -= sizeof(proof.storage_lmq_port);
+    lokimq::bt_dict bt_proof{
+      {"version", lokimq::bt_list{{proof.snode_version[0], proof.snode_version[1], proof.snode_version[2]}}},
+      {"pubkey",  tools::view_guts(proof.pubkey)},
```

Comment:
Starting in 8.x, lokid only generates the ed25519 key (and we use it for both pubkey_ed25519 *and* pubkey).  We can use that here: *if* the two pubkeys are the same then we can leave out the "pubkey" entry here, and the receiving side can notice that "pubkey" is missing and use "pubkey_ed25519" for both.

Similarly when signing: if the pubkeys are the same then we can only do the ed25519 signing, and the receiving entity will know that that's okay when "pubkey" is omitted.